cache/generic.go

```diff
@@ -60,38 +60,29 @@ func (c *Cache[K, V]) Get(ctx context.Context, key K) (V, error) {
 	return item.Value(), nil
 }
 
-type fetcher[K comparable, V any] struct {
-	item *ttlcache.Item[K, V]
-	err  error
-	done chan struct{}
-}
-
-func (c *Cache[K, V]) fetchAndSet(ctx context.Context, key K) (*ttlcache.Item[K, V], error) {
-	// try to set the fetcher for this key
-	actual, alreadyExists := c.fetchers.LoadOrStore(key, &fetcher[K, V]{done: make(chan struct{})})
-	fetcher := actual.(*fetcher[K, V])
-
-	if alreadyExists {
-		// wait for other fetcher routine to do the fetch
-		<-fetcher.done
-	} else {
-		defer func() {
-			c.fetchers.Delete(key)
-			close(fetcher.done)
-		}()
-
+func (c *Cache[K, V]) fetchAndSetSynced(ctx context.Context, key K) (*ttlcache.Item[K, V], error) {
+	ii, err, _ := c.fetchSync.Do(string(key), func() (any, error) {
 		// there's always a chance a different thread completed a fetch before we got here
 		// so check again now that we have a lock for the key
-		if item := c.cache.Get(key); item != nil {
-			fetcher.item, fetcher.err = item, nil
-		} else {
-			val, err := c.fetch(ctx, key)
-			if err != nil {
-				fetcher.err = err
-			} else {
-				fetcher.item = c.cache.Set(key, val, ttlcache.DefaultTTL)
-			}
+		item := c.cache.Get(key)
+		if item != nil {
+			return item, nil
```

Comment:
I can only get a thread here if I add a small random delay in `Get` between the cache lookup and the call to fetch and set, e.g. `time.Sleep(time.Duration(random.IntN(100)) * time.Millisecond)`